ApiImplementationGenerator.visitInvokeMethod does not properly handle parameters that occupy 2 slots on the stack

### Stacktrace

```
[23:35:41] [Server thread/ERROR]: Error occurred while enabling Chestshop v1.0 (Is it up to date?)
dev.derklaro.aerogel.AerogelException: Unable to construct requested value constructor ChestShopStorage((Object[])Object)
Caused by: eu.cloudnetservice.driver.network.rpc.exception.ClassCreationException: Unable to generate api class implementation for class net.tynopia.objects.interfaces.CitybuildSync
	at eu.cloudnetservice.driver.network.rpc.defaults.generation.ApiImplementationGenerator.generateApiImplementation(ApiImplementationGenerator.java:259) ~[wrapper.jar:4.0.0-RC7-ffd31747]
	at eu.cloudnetservice.driver.network.rpc.defaults.DefaultRPCFactory.lambda$generateRPCBasedApi$0(DefaultRPCFactory.java:122) ~[wrapper.jar:4.0.0-RC7-ffd31747]
	at com.github.benmanes.caffeine.cache.UnboundedLocalCache.lambda$computeIfAbsent$2(UnboundedLocalCache.java:289) ~[caffeine-3.1.2.jar:?]
	at java.util.concurrent.ConcurrentHashMap.computeIfAbsent(ConcurrentHashMap.java:1708) ~[?:?]
	at com.github.benmanes.caffeine.cache.UnboundedLocalCache.computeIfAbsent(UnboundedLocalCache.java:285) ~[caffeine-3.1.2.jar:?]
	at com.github.benmanes.caffeine.cache.LocalCache.computeIfAbsent(LocalCache.java:106) ~[caffeine-3.1.2.jar:?]
	at com.github.benmanes.caffeine.cache.LocalManualCache.get(LocalManualCache.java:62) ~[caffeine-3.1.2.jar:?]
	at eu.cloudnetservice.driver.network.rpc.defaults.DefaultRPCFactory.generateRPCBasedApi(DefaultRPCFactory.java:120) ~[wrapper.jar:4.0.0-RC7-ffd31747]
	at net.tynopia.Module.injectPlugin(Module.java:361) ~[APILibrary-1.0-SNAPSHOT.jar:?]
	at net.tynopia.chestshop.ChestShopStorage.<init>(ChestShopStorage.java:140) ~[Chestshop-1.0-SNAPSHOT.jar:?]
	at dev.derklaro.aerogel.internal.util.MethodHandleUtil.invokeConstructor(MethodHandleUtil.java:88) ~[aerogel-2.0.0.jar:?]
	at dev.derklaro.aerogel.internal.binding.constructors.ConstructingBindingConstructor.lambda$constructProvider$0(ConstructingBindingConstructor.java:92) ~[aerogel-2.0.0.jar:?]
	at dev.derklaro.aerogel.internal.provider.FunctionalContextualProvider.get(FunctionalContextualProvider.java:74) ~[aerogel-2.0.0.jar:?]
	at dev.derklaro.aerogel.internal.context.DefaultInjectionContext.resolveInstance(DefaultInjectionContext.java:388) ~[aerogel-2.0.0.jar:?]
	at dev.derklaro.aerogel.internal.context.util.ContextInstanceResolveHelper.resolveInstanceAndRemoveContext(ContextInstanceResolveHelper.java:95) ~[aerogel-2.0.0.jar:?]
	at dev.derklaro.aerogel.internal.context.util.ContextInstanceResolveHelper.resolveInstance(ContextInstanceResolveHelper.java:79) ~[aerogel-2.0.0.jar:?]
	at dev.derklaro.aerogel.internal.context.util.ContextInstanceResolveHelper.resolveInstance(ContextInstanceResolveHelper.java:63) ~[aerogel-2.0.0.jar:?]
	at dev.derklaro.aerogel.internal.DefaultInjector.instance(DefaultInjector.java:139) ~[aerogel-2.0.0.jar:?]
	at dev.derklaro.aerogel.internal.DefaultInjector.instance(DefaultInjector.java:121) ~[aerogel-2.0.0.jar:?]
	at net.tynopia.chestshop.ChestShopPlugin.onEnable(ChestShopPlugin.java:21) ~[Chestshop-1.0-SNAPSHOT.jar:?]
	at org.bukkit.plugin.java.JavaPlugin.setEnabled(JavaPlugin.java:264) ~[MultiPaper-API-1.19.3-R0.1-SNAPSHOT.jar:?]
	at org.bukkit.plugin.java.JavaPluginLoader.enablePlugin(JavaPluginLoader.java:372) ~[MultiPaper-API-1.19.3-R0.1-SNAPSHOT.jar:?]
	at org.bukkit.plugin.SimplePluginManager.enablePlugin(SimplePluginManager.java:550) ~[MultiPaper-API-1.19.3-R0.1-SNAPSHOT.jar:?]
	at org.bukkit.craftbukkit.v1_19_R2.CraftServer.enablePlugin(CraftServer.java:667) ~[multipaper-1.19.3.jar:git-MultiPaper-"0e5a6cf"]
	at org.bukkit.craftbukkit.v1_19_R2.CraftServer.enablePlugins(CraftServer.java:580) ~[multipaper-1.19.3.jar:git-MultiPaper-"0e5a6cf"]
	at net.minecraft.server.MinecraftServer.loadWorld0(MinecraftServer.java:644) ~[multipaper-1.19.3.jar:git-MultiPaper-"0e5a6cf"]
	at net.minecraft.server.MinecraftServer.loadLevel(MinecraftServer.java:443) ~[multipaper-1.19.3.jar:git-MultiPaper-"0e5a6cf"]
	at net.minecraft.server.dedicated.DedicatedServer.initServer(DedicatedServer.java:342) ~[multipaper-1.19.3.jar:git-MultiPaper-"0e5a6cf"]
	at net.minecraft.server.MinecraftServer.runServer(MinecraftServer.java:1123) ~[multipaper-1.19.3.jar:git-MultiPaper-"0e5a6cf"]
	at net.minecraft.server.MinecraftServer.lambda$spin$0(MinecraftServer.java:325) ~[multipaper-1.19.3.jar:git-MultiPaper-"0e5a6cf"]
	at java.lang.Thread.run(Thread.java:1589) ~[?:?]
Caused by: java.lang.IllegalStateException: Exception defining class net/tynopia/objects/interfaces/CitybuildSync$Impl_G0KSCP2Y9D
	at eu.cloudnetservice.driver.util.define.LookupClassDefiner.defineClass(LookupClassDefiner.java:62) ~[wrapper.jar:4.0.0-RC7-ffd31747]
	at eu.cloudnetservice.driver.network.rpc.defaults.generation.ApiImplementationGenerator.generateApiImplementation(ApiImplementationGenerator.java:243) ~[wrapper.jar:4.0.0-RC7-ffd31747]
	at eu.cloudnetservice.driver.network.rpc.defaults.DefaultRPCFactory.lambda$generateRPCBasedApi$0(DefaultRPCFactory.java:122) ~[wrapper.jar:4.0.0-RC7-ffd31747]
	at com.github.benmanes.caffeine.cache.UnboundedLocalCache.lambda$computeIfAbsent$2(UnboundedLocalCache.java:289) ~[caffeine-3.1.2.jar:?]
	at java.util.concurrent.ConcurrentHashMap.computeIfAbsent(ConcurrentHashMap.java:1708) ~[?:?]
	at com.github.benmanes.caffeine.cache.UnboundedLocalCache.computeIfAbsent(UnboundedLocalCache.java:285) ~[caffeine-3.1.2.jar:?]
	at com.github.benmanes.caffeine.cache.LocalCache.computeIfAbsent(LocalCache.java:106) ~[caffeine-3.1.2.jar:?]
	at com.github.benmanes.caffeine.cache.LocalManualCache.get(LocalManualCache.java:62) ~[caffeine-3.1.2.jar:?]
	at eu.cloudnetservice.driver.network.rpc.defaults.DefaultRPCFactory.generateRPCBasedApi(DefaultRPCFactory.java:120) ~[wrapper.jar:4.0.0-RC7-ffd31747]
	at net.tynopia.Module.injectPlugin(Module.java:361) ~[APILibrary-1.0-SNAPSHOT.jar:?]
	at net.tynopia.chestshop.ChestShopStorage.<init>(ChestShopStorage.java:140) ~[Chestshop-1.0-SNAPSHOT.jar:?]
	at dev.derklaro.aerogel.internal.util.MethodHandleUtil.invokeConstructor(MethodHandleUtil.java:88) ~[aerogel-2.0.0.jar:?]
	at dev.derklaro.aerogel.internal.binding.constructors.ConstructingBindingConstructor.lambda$constructProvider$0(ConstructingBindingConstructor.java:92) ~[aerogel-2.0.0.jar:?]
	at dev.derklaro.aerogel.internal.provider.FunctionalContextualProvider.get(FunctionalContextualProvider.java:74) ~[aerogel-2.0.0.jar:?]
	at dev.derklaro.aerogel.internal.context.DefaultInjectionContext.resolveInstance(DefaultInjectionContext.java:388) ~[aerogel-2.0.0.jar:?]
	at dev.derklaro.aerogel.internal.context.util.ContextInstanceResolveHelper.resolveInstanceAndRemoveContext(ContextInstanceResolveHelper.java:95) ~[aerogel-2.0.0.jar:?]
	at dev.derklaro.aerogel.internal.context.util.ContextInstanceResolveHelper.resolveInstance(ContextInstanceResolveHelper.java:79) ~[aerogel-2.0.0.jar:?]
	at dev.derklaro.aerogel.internal.context.util.ContextInstanceResolveHelper.resolveInstance(ContextInstanceResolveHelper.java:63) ~[aerogel-2.0.0.jar:?]
	at dev.derklaro.aerogel.internal.DefaultInjector.instance(DefaultInjector.java:139) ~[aerogel-2.0.0.jar:?]
	at dev.derklaro.aerogel.internal.DefaultInjector.instance(DefaultInjector.java:121) ~[aerogel-2.0.0.jar:?]
	at net.tynopia.chestshop.ChestShopPlugin.onEnable(ChestShopPlugin.java:21) ~[Chestshop-1.0-SNAPSHOT.jar:?]
	at org.bukkit.plugin.java.JavaPlugin.setEnabled(JavaPlugin.java:264) ~[MultiPaper-API-1.19.3-R0.1-SNAPSHOT.jar:?]
	at org.bukkit.plugin.java.JavaPluginLoader.enablePlugin(JavaPluginLoader.java:372) ~[MultiPaper-API-1.19.3-R0.1-SNAPSHOT.jar:?]
	at org.bukkit.plugin.SimplePluginManager.enablePlugin(SimplePluginManager.java:550) ~[MultiPaper-API-1.19.3-R0.1-SNAPSHOT.jar:?]
	at org.bukkit.craftbukkit.v1_19_R2.CraftServer.enablePlugin(CraftServer.java:667) ~[multipaper-1.19.3.jar:git-MultiPaper-"0e5a6cf"]
	at org.bukkit.craftbukkit.v1_19_R2.CraftServer.enablePlugins(CraftServer.java:580) ~[multipaper-1.19.3.jar:git-MultiPaper-"0e5a6cf"]
	at net.minecraft.server.MinecraftServer.loadWorld0(MinecraftServer.java:644) ~[multipaper-1.19.3.jar:git-MultiPaper-"0e5a6cf"]
	at net.minecraft.server.MinecraftServer.loadLevel(MinecraftServer.java:443) ~[multipaper-1.19.3.jar:git-MultiPaper-"0e5a6cf"]
	at net.minecraft.server.dedicated.DedicatedServer.initServer(DedicatedServer.java:342) ~[multipaper-1.19.3.jar:git-MultiPaper-"0e5a6cf"]
	at net.minecraft.server.MinecraftServer.runServer(MinecraftServer.java:1123) ~[multipaper-1.19.3.jar:git-MultiPaper-"0e5a6cf"]
	at net.minecraft.server.MinecraftServer.lambda$spin$0(MinecraftServer.java:325) ~[multipaper-1.19.3.jar:git-MultiPaper-"0e5a6cf"]
	at java.lang.Thread.run(Thread.java:1589) ~[?:?]
Caused by: java.lang.VerifyError: Bad local variable type
Exception Details:
  Location:
    net/tynopia/objects/interfaces/CitybuildSync$Impl_G0KSCP2Y9D+0x0000000802799400.setChestShop(JIIILjava/lang/String;Ljava/lang/String;)I @20: iload_2
  Reason:
    Type long_2nd (current frame, locals[2]) is not assignable to integer
  Current Frame:
    bci: @20
    flags: { }
    locals: { 'net/tynopia/objects/interfaces/CitybuildSync$Impl_G0KSCP2Y9D+0x0000000802799400', long, long_2nd, integer, integer, integer, 'java/lang/String', 'java/lang/String' }
    stack: { 'eu/cloudnetservice/driver/network/rpc/RPCSender', 'java/lang/String', '[Ljava/lang/Object;', '[Ljava/lang/Object;', integer }
  Bytecode:
    0000000: 2ab4 0012 1256 1006 bd00 0459 031f b800
    0000010: 5b53 5904 1cb8 004d 5359 051d b800 4d53
    0000020: 5906 1504 b800 4d53 5907 1905 5359 0819
    0000030: 0653 b900 1d03 00b9 0023 0100 c000 2ab6
    0000040: 002e ac                                

	at java.lang.ClassLoader.defineClass0(Native Method) ~[?:?]
	at java.lang.System$2.defineClass(System.java:2371) ~[?:?]
	at java.lang.invoke.MethodHandles$Lookup$ClassDefiner.defineClass(MethodHandles.java:2449) ~[?:?]
	at java.lang.invoke.MethodHandles$Lookup$ClassDefiner.defineClassAsLookup(MethodHandles.java:2430) ~[?:?]
	at java.lang.invoke.MethodHandles$Lookup.defineHiddenClass(MethodHandles.java:2136) ~[?:?]
	at eu.cloudnetservice.driver.util.define.LookupClassDefiner.defineClass(LookupClassDefiner.java:59) ~[wrapper.jar:4.0.0-RC7-ffd31747]
	at eu.cloudnetservice.driver.network.rpc.defaults.generation.ApiImplementationGenerator.generateApiImplementation(ApiImplementationGenerator.java:243) ~[wrapper.jar:4.0.0-RC7-ffd31747]
	at eu.cloudnetservice.driver.network.rpc.defaults.DefaultRPCFactory.lambda$generateRPCBasedApi$0(DefaultRPCFactory.java:122) ~[wrapper.jar:4.0.0-RC7-ffd31747]
	at com.github.benmanes.caffeine.cache.UnboundedLocalCache.lambda$computeIfAbsent$2(UnboundedLocalCache.java:289) ~[caffeine-3.1.2.jar:?]
	at java.util.concurrent.ConcurrentHashMap.computeIfAbsent(ConcurrentHashMap.java:1708) ~[?:?]
	at com.github.benmanes.caffeine.cache.UnboundedLocalCache.computeIfAbsent(UnboundedLocalCache.java:285) ~[caffeine-3.1.2.jar:?]
	at com.github.benmanes.caffeine.cache.LocalCache.computeIfAbsent(LocalCache.java:106) ~[caffeine-3.1.2.jar:?]
	at com.github.benmanes.caffeine.cache.LocalManualCache.get(LocalManualCache.java:62) ~[caffeine-3.1.2.jar:?]
	at eu.cloudnetservice.driver.network.rpc.defaults.DefaultRPCFactory.generateRPCBasedApi(DefaultRPCFactory.java:120) ~[wrapper.jar:4.0.0-RC7-ffd31747]
	at net.tynopia.Module.injectPlugin(Module.java:361) ~[APILibrary-1.0-SNAPSHOT.jar:?]
	at net.tynopia.chestshop.ChestShopStorage.<init>(ChestShopStorage.java:140) ~[Chestshop-1.0-SNAPSHOT.jar:?]
	at dev.derklaro.aerogel.internal.util.MethodHandleUtil.invokeConstructor(MethodHandleUtil.java:88) ~[aerogel-2.0.0.jar:?]
	at dev.derklaro.aerogel.internal.binding.constructors.ConstructingBindingConstructor.lambda$constructProvider$0(ConstructingBindingConstructor.java:92) ~[aerogel-2.0.0.jar:?]
	at dev.derklaro.aerogel.internal.provider.FunctionalContextualProvider.get(FunctionalContextualProvider.java:74) ~[aerogel-2.0.0.jar:?]
	at dev.derklaro.aerogel.internal.context.DefaultInjectionContext.resolveInstance(DefaultInjectionContext.java:388) ~[aerogel-2.0.0.jar:?]
	at dev.derklaro.aerogel.internal.context.util.ContextInstanceResolveHelper.resolveInstanceAndRemoveContext(ContextInstanceResolveHelper.java:95) ~[aerogel-2.0.0.jar:?]
	at dev.derklaro.aerogel.internal.context.util.ContextInstanceResolveHelper.resolveInstance(ContextInstanceResolveHelper.java:79) ~[aerogel-2.0.0.jar:?]
	at dev.derklaro.aerogel.internal.context.util.ContextInstanceResolveHelper.resolveInstance(ContextInstanceResolveHelper.java:63) ~[aerogel-2.0.0.jar:?]
	at dev.derklaro.aerogel.internal.DefaultInjector.instance(DefaultInjector.java:139) ~[aerogel-2.0.0.jar:?]
	at dev.derklaro.aerogel.internal.DefaultInjector.instance(DefaultInjector.java:121) ~[aerogel-2.0.0.jar:?]
	at net.tynopia.chestshop.ChestShopPlugin.onEnable(ChestShopPlugin.java:21) ~[Chestshop-1.0-SNAPSHOT.jar:?]
	at org.bukkit.plugin.java.JavaPlugin.setEnabled(JavaPlugin.java:264) ~[MultiPaper-API-1.19.3-R0.1-SNAPSHOT.jar:?]
	at org.bukkit.plugin.java.JavaPluginLoader.enablePlugin(JavaPluginLoader.java:372) ~[MultiPaper-API-1.19.3-R0.1-SNAPSHOT.jar:?]
	at org.bukkit.plugin.SimplePluginManager.enablePlugin(SimplePluginManager.java:550) ~[MultiPaper-API-1.19.3-R0.1-SNAPSHOT.jar:?]
	at org.bukkit.craftbukkit.v1_19_R2.CraftServer.enablePlugin(CraftServer.java:667) ~[multipaper-1.19.3.jar:git-MultiPaper-"0e5a6cf"]
	at org.bukkit.craftbukkit.v1_19_R2.CraftServer.enablePlugins(CraftServer.java:580) ~[multipaper-1.19.3.jar:git-MultiPaper-"0e5a6cf"]
	at net.minecraft.server.MinecraftServer.loadWorld0(MinecraftServer.java:644) ~[multipaper-1.19.3.jar:git-MultiPaper-"0e5a6cf"]
	at net.minecraft.server.MinecraftServer.loadLevel(MinecraftServer.java:443) ~[multipaper-1.19.3.jar:git-MultiPaper-"0e5a6cf"]
	at net.minecraft.server.dedicated.DedicatedServer.initServer(DedicatedServer.java:342) ~[multipaper-1.19.3.jar:git-MultiPaper-"0e5a6cf"]
	at net.minecraft.server.MinecraftServer.runServer(MinecraftServer.java:1123) ~[multipaper-1.19.3.jar:git-MultiPaper-"0e5a6cf"]
	at net.minecraft.server.MinecraftServer.lambda$spin$0(MinecraftServer.java:325) ~[multipaper-1.19.3.jar:git-MultiPaper-"0e5a6cf"]
	at java.lang.Thread.run(Thread.java:1589) ~[?:?]
[23:35:41] [Server thread/INFO]: [Chestshop] Disabling Chestshop v1.0
```


### Actions to reproduce

Create an RPC method that has at least one `long` or `double` argument and another argument following that.

### CloudNet version

4.0.0-RC7-ffd31747

### Other

Example method signature: `int setChestShop(UUID var1, long var2, int var4, int var5, int var6, String var7, String var8)`  
This is in reference to the following Discord discussion: https://discord.com/channels/325362837184577536/326369842229805076/1074791477223358495

In essence the code is expecting each parameter to only ever occupy one slot on the stack. Given that longs and doubles use two slots on the stack, the following code tries to read the second half of the long/double as the next parameter: https://github.com/CloudNetService/CloudNet-v3/blob/554d56d5a3a34df3abd827d4ab8a9d8e14afc8e9/driver/src/main/java/eu/cloudnetservice/driver/network/rpc/defaults/generation/ApiImplementationGenerator.java#L284-L295

In order to fix this, the two known types of `long` and `double` (both explicitly not boxed) need to cause the loop to skip forward one slot.
In any case, the methods of ASM work correctly loading a long and then an int: https://asm.ow2.io/javadoc/org/objectweb/asm/Type.html#getOpcode(int).

### Issue uniqueness

- [X] Yes, this issue is unique. There are no similar issues.

--- a/driver/src/main/java/eu/cloudnetservice/driver/network/rpc/defaults/generation/ApiImplementationGenerator.java
+++ b/driver/src/main/java/eu/cloudnetservice/driver/network/rpc/defaults/generation/ApiImplementationGenerator.java
@@ -25,7 +25,6 @@ import static org.objectweb.asm.Opcodes.ANEWARRAY;
 import static org.objectweb.asm.Opcodes.CHECKCAST;
 import static org.objectweb.asm.Opcodes.DUP;
 import static org.objectweb.asm.Opcodes.GETFIELD;
-import static org.objectweb.asm.Opcodes.ILOAD;
 import static org.objectweb.asm.Opcodes.INVOKEINTERFACE;
 import static org.objectweb.asm.Opcodes.INVOKESPECIAL;
 import static org.objectweb.asm.Opcodes.IRETURN;
@@ -50,6 +49,7 @@ import eu.cloudnetservice.driver.network.rpc.exception.ClassCreationException;
 import eu.cloudnetservice.driver.network.rpc.generation.GenerationContext;
 import eu.cloudnetservice.driver.network.rpc.generation.InstanceFactory;
 import eu.cloudnetservice.driver.util.asm.AsmHelper;
+import eu.cloudnetservice.driver.util.asm.StackIndexHelper;
 import eu.cloudnetservice.driver.util.define.ClassDefiners;
 import jakarta.inject.Inject;
 import java.lang.reflect.Constructor;
@@ -183,9 +183,9 @@ public final class ApiImplementationGenerator {
         mv.visitMethodInsn(INVOKESPECIAL, superName, "<init>", "()V", false);
       } else {
         // load all parameters to the stack
+        var stackHelper = StackIndexHelper.create(3); // 0 = this, 1 = sender, 2 = channel supplier
         for (int i = 0; i < targetConstructorArgs.size(); i++) {
           var argumentType = targetConstructorArgs.get(i);
-
           if (i == 0 && argumentType.equals(RPCSender.class)) {
             // argument is the provided sender
             mv.visitVarInsn(ALOAD, 1);
@@ -193,8 +193,8 @@ public final class ApiImplementationGenerator {
             // argument is the provided channel supplier
             mv.visitVarInsn(ALOAD, 2);
           } else {
-            var type = Type.getType(argumentType);
-            mv.visitVarInsn(type.getOpcode(ILOAD), 2 + i); // 0 = this, 1 = sender, 2 = channel supplier
+            // argument is provided by the constructor
+            stackHelper.load(mv, argumentType);
           }
         }
 
@@ -280,22 +280,26 @@ public final class ApiImplementationGenerator {
     var params = method.getParameterTypes();
     AsmHelper.pushInt(mv, params.length);
     mv.visitTypeInsn(ANEWARRAY, DEFAULT_SUPER);
+
     // add each parameter to the array
+    var stackHelper = StackIndexHelper.create().skipNext(); // skip "this"
     for (var i = 0; i < params.length; i++) {
-      var type = params[i];
-      var intType = Type.getType(type);
+      var paramType = params[i];
+
       // dup the array on the stack and push the target index we want to add the object to
       mv.visitInsn(DUP);
       AsmHelper.pushInt(mv, i);
-      // load the parameter
-      mv.visitVarInsn(intType.getOpcode(ILOAD), i + 1);
-      // we need to wrap a primitive type to the wrapper type
-      if (type.isPrimitive()) {
-        AsmHelper.primitiveToWrapper(mv, type);
+
+      // load the parameter & wrap it in a boxed type if needed
+      stackHelper.load(mv, paramType);
+      if (paramType.isPrimitive()) {
+        AsmHelper.primitiveToWrapper(mv, paramType);
       }
-      // store the element
+
+      // store the element in the array
       mv.visitInsn(AASTORE);
     }
+
     // invoke the send method
     mv.visitMethodInsn(INVOKEINTERFACE, SENDER_TYPE, "invokeMethod", INVOKE_METHOD_DESC, true);
   }

--- a/driver/src/main/java/eu/cloudnetservice/driver/network/rpc/defaults/generation/ChainedApiImplementationGenerator.java
+++ b/driver/src/main/java/eu/cloudnetservice/driver/network/rpc/defaults/generation/ChainedApiImplementationGenerator.java
@@ -28,7 +28,6 @@ import static org.objectweb.asm.Opcodes.ACC_PRIVATE;
 import static org.objectweb.asm.Opcodes.ACC_PUBLIC;
 import static org.objectweb.asm.Opcodes.ALOAD;
 import static org.objectweb.asm.Opcodes.GETFIELD;
-import static org.objectweb.asm.Opcodes.ILOAD;
 import static org.objectweb.asm.Opcodes.INVOKEINTERFACE;
 import static org.objectweb.asm.Opcodes.INVOKESPECIAL;
 import static org.objectweb.asm.Opcodes.PUTFIELD;
@@ -46,6 +45,7 @@ import eu.cloudnetservice.driver.network.rpc.RPCSender;
 import eu.cloudnetservice.driver.network.rpc.exception.ClassCreationException;
 import eu.cloudnetservice.driver.network.rpc.generation.ChainInstanceFactory;
 import eu.cloudnetservice.driver.network.rpc.generation.GenerationContext;
+import eu.cloudnetservice.driver.util.asm.StackIndexHelper;
 import eu.cloudnetservice.driver.util.define.ClassDefiners;
 import java.lang.reflect.Constructor;
 import java.util.Arrays;
@@ -140,13 +140,13 @@ public final class ChainedApiImplementationGenerator {
         mv.visitVarInsn(ALOAD, 3);
         mv.visitFieldInsn(PUTFIELD, className, "channelSupplier", SUPPLIER_DESC);
 
-        // call the super "<init>" method
+        // load the arguments for the super constructor to the stack
         mv.visitVarInsn(ALOAD, 0);
-        for (var i = 0; i < types.size(); i++) {
-          // loads and passes the correct arguments to the super constructor
-          var type = Type.getType(types.get(i));
-          mv.visitVarInsn(type.getOpcode(ILOAD), 4 + i);
+        var stackHelper = StackIndexHelper.create(4); // 0 = this, 1-3 are taken by required parameters
+        for (var type : types) {
+          stackHelper.load(mv, type);
         }
+        // visit the super constructor call
         mv.visitMethodInsn(INVOKESPECIAL, superName, "<init>", '(' + superDesc + ")V", false);
 
         // finish

--- a/driver/src/main/java/eu/cloudnetservice/driver/util/asm/StackIndexHelper.java
+++ b/driver/src/main/java/eu/cloudnetservice/driver/util/asm/StackIndexHelper.java
@@ -1,0 +1,146 @@
+/*
+ * Copyright 2019-2023 CloudNetService team & contributors
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package eu.cloudnetservice.driver.util.asm;
+
+import lombok.NonNull;
+import org.jetbrains.annotations.ApiStatus;
+import org.objectweb.asm.MethodVisitor;
+import org.objectweb.asm.Opcodes;
+import org.objectweb.asm.Type;
+
+/**
+ * A small helper class that helps keeping track of bytecode stack indices while pushing/loading from it.
+ *
+ * @since 4.0
+ */
+@ApiStatus.Internal
+public final class StackIndexHelper {
+
+  private int nextVarIndex;
+
+  /**
+   * Constructs a new stack index helper instance.
+   *
+   * @param initialVarIndex the initial var index for the first element to load/store from/to the stack.
+   */
+  private StackIndexHelper(int initialVarIndex) {
+    this.nextVarIndex = initialVarIndex;
+  }
+
+  /**
+   * Creates a new stack index helper which uses 0 as the initial index.
+   *
+   * @return a stack index helper that uses 0 as the initial index.
+   */
+  public static @NonNull StackIndexHelper create() {
+    return create(0);
+  }
+
+  /**
+   * Constructs a new stack index helper with the given initial var index.
+   *
+   * @param initialVarIndex the initial var index for the first element to load/store from/to the stack.
+   * @return a stack index helper that uses the given initial var index.
+   */
+  public static @NonNull StackIndexHelper create(int initialVarIndex) {
+    return new StackIndexHelper(initialVarIndex);
+  }
+
+  /**
+   * Skips the next (one) element of the current stack. This method implies that the next element has a size of one.
+   *
+   * @return this index helper, for chaining.
+   */
+  public @NonNull StackIndexHelper skipNext() {
+    return this.skip(1);
+  }
+
+  /**
+   * Skips the given amount of indices and uses the resulting value as the next var index to read/write from/to.
+   *
+   * @param depth the stack depth to skip.
+   * @return this index helper, for chaining.
+   */
+  public @NonNull StackIndexHelper skip(int depth) {
+    this.nextVarIndex += depth;
+    return this;
+  }
+
+  /**
+   * Stores an element of the given type on the stack using the appropriate opcode based on the given type and skips to
+   * the next var index based on the size of the type.
+   *
+   * @param methodVisitor the method visitor to notify about the store operation.
+   * @param type          the type to store on the stack.
+   * @return this index helper, for chaining.
+   * @throws NullPointerException if the given visitor or type is null.
+   */
+  public @NonNull StackIndexHelper push(@NonNull MethodVisitor methodVisitor, @NonNull Class<?> type) {
+    var internalType = Type.getType(type);
+    return this.push(methodVisitor, internalType);
+  }
+
+  /**
+   * Stores an element of the given type on the stack using the appropriate opcode based on the given type and skips to
+   * the next var index based on the size of the type.
+   *
+   * @param methodVisitor the method visitor to notify about the store operation.
+   * @param type          the type to store on the stack.
+   * @return this index helper, for chaining.
+   * @throws NullPointerException if the given visitor or type is null.
+   */
+  public @NonNull StackIndexHelper push(@NonNull MethodVisitor methodVisitor, @NonNull Type type) {
+    // store the element on the stack using the associated opcode
+    var opcode = type.getOpcode(Opcodes.ISTORE);
+    methodVisitor.visitVarInsn(opcode, this.nextVarIndex);
+
+    // skip to the next var index based on the size of the type
+    return this.skip(type.getSize());
+  }
+
+  /**
+   * Loads an element of the given type onto the stack using the appropriate opcode based on the given type and skips to
+   * the next var index based on the size of the type.
+   *
+   * @param methodVisitor the method visitor to notify about the load operation.
+   * @param type          the type to load from the stack.
+   * @return this index helper, for chaining.
+   * @throws NullPointerException if the given visitor or type is null.
+   */
+  public @NonNull StackIndexHelper load(@NonNull MethodVisitor methodVisitor, @NonNull Class<?> type) {
+    var internalType = Type.getType(type);
+    return this.load(methodVisitor, internalType);
+  }
+
+  /**
+   * Loads an element of the given type onto the stack using the appropriate opcode based on the given type and skips to
+   * the next var index based on the size of the type.
+   *
+   * @param methodVisitor the method visitor to notify about the load operation.
+   * @param type          the type to load from the stack.
+   * @return this index helper, for chaining.
+   * @throws NullPointerException if the given visitor or type is null.
+   */
+  public @NonNull StackIndexHelper load(@NonNull MethodVisitor methodVisitor, @NonNull Type type) {
+    // load the element from the stack using the associated opcode
+    var opcode = type.getOpcode(Opcodes.ILOAD);
+    methodVisitor.visitVarInsn(opcode, this.nextVarIndex);
+
+    // skip to the next var index based on the size of the type
+    return this.skip(type.getSize());
+  }
+}

--- a/driver/src/test/java/eu/cloudnetservice/driver/network/rpc/generation/api/RPCImplementationGeneratorTest.java
+++ b/driver/src/test/java/eu/cloudnetservice/driver/network/rpc/generation/api/RPCImplementationGeneratorTest.java
@@ -92,4 +92,14 @@ public class RPCImplementationGeneratorTest {
     Assertions.assertInstanceOf(SenderNeedingManagement.class, management);
     Assertions.assertSame(((SenderNeedingManagement) management).sender, sender);
   }
+
+  @Test
+  void testGenerationOfMethods() {
+    var testInstance = ApiImplementationGenerator.generateApiImplementation(
+      TestRPCParameters.class,
+      GenerationContext.forClass(TestRPCParameters.class).implementAllMethods(true).build(),
+      Mockito.mock(RPCSender.class)
+    ).newInstance(1, 2L, "World", new int[0], new long[]{1L}, new String[]{"Hello"});
+    Assertions.assertNotNull(testInstance);
+  }
 }

--- a/driver/src/test/java/eu/cloudnetservice/driver/network/rpc/generation/api/TestRPCParameters.java
+++ b/driver/src/test/java/eu/cloudnetservice/driver/network/rpc/generation/api/TestRPCParameters.java
@@ -1,0 +1,75 @@
+/*
+ * Copyright 2019-2023 CloudNetService team & contributors
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package eu.cloudnetservice.driver.network.rpc.generation.api;
+
+public abstract class TestRPCParameters {
+
+  public TestRPCParameters(int a, long b, String c, int[] d, long[] e, String[] f) {
+  }
+
+  public abstract void testVoid();
+
+  public abstract void testString(String string);
+
+  public abstract void testInt(int i);
+
+  public abstract void testLong(long l);
+
+  public abstract void testFloat(float f);
+
+  public abstract void testDouble(double d);
+
+  public abstract void testBoolean(boolean b);
+
+  public abstract void testChar(char c);
+
+  public abstract void testByte(byte b);
+
+  public abstract void testShort(short s);
+
+  public abstract void testObject(Object o);
+
+  public abstract void testStringArray(String[] strings);
+
+  public abstract void testIntArray(int[] ints);
+
+  public abstract void testLongArray(long[] longs);
+
+  public abstract void testFloatArray(float[] floats);
+
+  public abstract void testDoubleArray(double[] doubles);
+
+  public abstract void testBooleanArray(boolean[] booleans);
+
+  public abstract void testCharArray(char[] chars);
+
+  public abstract void testByteArray(byte[] bytes);
+
+  public abstract void testShortArray(short[] shorts);
+
+  public abstract void testObjectArray(Object[] objects);
+
+  public abstract void testLongInt(long l, int i);
+
+  public abstract void testLongIntString(long l, int i, String s);
+
+  public abstract void testLongIntStringArray(long l, int i, String[] s);
+
+  public abstract void testLongIntStringObject(long l, int i, String s, Object o);
+
+  public abstract void testLongIntStringObjectArray(long l, int i, String s, Object[] o);
+}

